Extensions/BBText/bbtextruntimeobject-pixi-renderer.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * The PIXI.js renderer for the BBCode Text runtime object.
+ *
+ * @class BBTextRuntimeObjectPixiRenderer
+ * @constructor
+ * @param {gdjs.BBTextRuntimeObject} runtimeObject The object to render
+ * @param {gdjs.RuntimeScene} runtimeScene The gdjs.RuntimeScene in which the object is
+ */
+gdjs.BBTextRuntimeObjectPixiRenderer = function(runtimeObject, runtimeScene) {
+  this._object = runtimeObject;
+
+  // Load (or reset) the text
+  if (this._pixiObject === undefined) {
+    this._pixiObject = new MultiStyleText(runtimeObject._text, {
+      default: {
+        fontFamily: runtimeObject._fontFamily,
+        fontSize: `${runtimeObject._fontSize}px`,
+        fill: runtimeObject._color,
+        tagStyle: 'bbcode',
+        wordWrap: runtimeObject._wordWrap,
+        wordWrapWidth: runtimeObject._wrappingWidth,
+        align: runtimeObject._align,
+      },
+    });
+
+    this._object.hidden = !runtimeObject._visible;
+  } else {
+    this.updateText();
+    this.updateColor();
+    this.updateAlignment();
+    this.updateFontFamily();
+    this.updateFontSize();
+    this.updatePosition();
+  }
+
+  runtimeScene
+    .getLayer('')
+    .getRenderer()
+    .addRendererObject(this._pixiObject, runtimeObject.getZOrder());
+
+  // Set the anchor in the center, so that the object rotates around
+  // its center
+  this._pixiObject.anchor.x = 0.5;
+  this._pixiObject.anchor.y = 0.5;
+
+  this.updateText();
+  this.updatePosition();
+  this.updateAngle();
+  this.updateOpacity();
+  this.updateVisible();
+  this.ensureUpToDate();
+};
+
+gdjs.BBTextRuntimeObjectRenderer = gdjs.BBTextRuntimeObjectPixiRenderer;
+
+gdjs.BBTextRuntimeObjectPixiRenderer.prototype.getRendererObject = function() {
+  return this._pixiObject;
+};
+
+gdjs.BBTextRuntimeObjectPixiRenderer.prototype.ensureUpToDate = function() {
+  if (this._pixiObject._style.wordWrap !== this._object._wordWrap) {
+    this.updateWordWrap();
+  }
+  if (this._pixiObject._style.wordWrapWidth !== this._object._wrappingWidth) {
+    this.updateWrappingWidth();
+  }
+  if (
+    !this._textureWasValid &&
```

Comment:
This looks like some code remaining from the video object. If it's useful, then you must declare and set `_textureWasValid` to false in the constructor because:
- it's bad for performance to create properties at runtime. Always declare everything in your constructor, otherwise the JS engine must do weird tricks to accomodate your object adding new properties.
- otherwise, this will stay to true if the object is recycled, creating probably weird bug/glitchs.

Reply:
Verify that `this._pixiObject.texture.valid` is even something that exists for your Pixi object :)

Reply:
> Verify that `this._pixiObject.texture.valid` is even something that exists for your Pixi object :)

I think I can remove the entire ensureUptoDate function. You are right, if this texture.valid is not something that is a part of gd's extension api, it has so point of existing there

Reply:
So you know how in GD when you first drop an instance, it is dropped with width and height set to 0. This is a problem for my extension. If the width is zero and user hasnt resized it, what is dropped in the IDE and what is rendered in game have different wrapping width

Reply:
See if you can remove it entirely yes :) 

> So when that width value is initiated as 0, I need my extension to use the width boundaries from the ide renderer

The width is maybe 0, but in this case it's because customSize is false, so your object should not use this to set any width.

Basically:
- Is customSize set on initial instance? If yes, that means that the user resized the object in the IDE. Set the proper width and height as stored on the initial instance.
- If no, that means that the user put the object on the scene without resizing it. The object is expected to take it's default size. In the case of a text, this means that the width is as large as the longest line, there is no wrapping.

Reply:
Knowing this, this means that always calling setWrappingWidth in extraInitializationFromInitialInstance is suspect. See how it's done for the builtin Text object:

https://github.com/4ian/GDevelop/blob/83d0c4cc65d6a8a9635ef09285c4a7ebb6bd085b/Extensions/TextObject/textruntimeobject.js#L67-L72

Reply:
>      if ( initialInstanceData.customSize ) { 
>          this.setWrapping(true); 
>          this.setWrappingWidth(initialInstanceData.width); 
>      }

ah I see, in that case I will copy the approach

Reply:
That doesn't seem to resolve the problem :) not sure what to do with this

Reply:
Can you describe the problem precisely to see what could go wrong? As well as adding a few console.log() for variables (initialInstanceData.customSize, initialInstanceData.width, etc...) to get an idea.

Reply:
so the problem is basically reproducible by doing this:
1. create an object with a long text without new lines
2. drop an instance of it on the scene without changing anything
3. Playtest

what is rendered in the IDE is different from what is rendered in the game

Reason: Width is set to 0 by default

fix: every time you drop an instance of a bbtext object, you need to resize it, so its width value is not a zero

Reply:
> I will try to fix this tonight :)

Ok! Try to think about what's different between the instance rendered in the IDE and in the game. 

> The problem is having different rendering in the IDE from the game when width is set to 0

Remember that width should NOT even be read if customSize is false. This is not a value that you always read. If `customSize` if false, this value can be anything (-1, 0, 9999), your object should not use it and display like in the IDE.

Reply:
I found out the reason for the difference. The default wrapping width that the object is initiated with in the IDE (250) was different from the one when initiated by the runtime (0). 
In this case I feel that 0 is a really bad default that nobody would like to use for wrapping width, so I will keep it at 250, which will be overwritten when the instance has been resized by the user

Reply:
I noticed that GD has another quirk in general - with the ordinary text object too - if you drop a text object instance, it has 0 as the default width. If you tick on "custom size", the object suddenly gets resized to 0 width. Perhaps I am now nitpicking, but just an observation that this is present in other objects

Reply:
>  you tick on "custom size", the object suddenly gets resized to 0 width

Yes that could indeed be improved by setting the size (width and height) of the instance to the size reported by the instance renderer (which is the only thing to be able to report the current size of the instance). This would need some refactoring to give back this information to the properties grid though I think.

> The default wrapping width that the object is initiated with in the IDE (250) 

Do you know why this size is 250 by default in the IDE?